core/filelocation.cpp

```diff
@@ -0,0 +1,235 @@
+// SPDX-License-Identifier: GPL-2.0
+#include "core/filelocation.h"
+#include "core/pref.h"
+#include <QObject>
+#include <QFileInfo>
+#include <QDir>
+#include <utility>
+
+FileLocation currentFile;
+
+FileLocation::FileLocation() : type(NONE)
+{
+}
+
+bool isRemoteRepository(const QString &name)
+{
+	return name.startsWith("http://") || name.startsWith("https://") ||
+	       name.startsWith("ssh://") || name.startsWith("git://");
+}
+
+// For git repositories, remove "file://" and extract username, if any.
+// For local files or git repositories, do proper unicode encoding.
+void FileLocation::fixName()
+{
+	switch (type) {
+	default:
+	case NONE:
+		return;
+	case LOCAL_FILE:
+		name = QFile::encodeName(QDir::toNativeSeparators(name));
+		return;
+	case CLOUD_GIT:
+	case CLOUD_GIT_OFFLINE:
+	case GIT:
+		if (name.startsWith("file://", Qt::CaseInsensitive)) {
+			// Treat "file://..." URLs as local repositories.
+			name = name.mid(7);
+			name = QFile::encodeName(QDir::toNativeSeparators(name));
+		} else if(!isRemoteRepository(name)) {
```

Comment:
whitespace: we always want a space between "if" and the "(" : if (...  
You do that some times, but not others